recorderproxy/recorderproxy.go

```diff
@@ -0,0 +1,448 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package recorderproxy
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/nlnwa/veidemann-api-go/browsercontroller/v1"
+	"github.com/nlnwa/veidemann-api-go/commons/v1"
+	"github.com/nlnwa/veidemann-api-go/config/v1"
+	"github.com/nlnwa/veidemann-api-go/contentwriter/v1"
+	"github.com/nlnwa/veidemann-api-go/dnsresolver/v1"
+	"github.com/nlnwa/veidemann-api-go/frontier/v1"
+	"github.com/spf13/viper"
+	"io"
+	"io/ioutil"
+	"regexp"
+
+	//log "github.com/sirupsen/logrus"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	ENCODING         = "Accept-Encoding"
+	EXECUTION_ID     = "veidemann_eid"
+	JOB_EXECUTION_ID = "veidemann_jeid"
+	COLLECTION_ID    = "veidemann_cid"
+)
+
+const (
+	ContentTypeText = "text/plain"
+	ContentTypeHtml = "text/html"
+)
+
+var proxyCount int32
+
+type RecorderProxy struct {
+	id                int32
+	addr              string
+	conn              Connections
+	ConnectionTimeout time.Duration
+
+	// session variable must be aligned in i386
+	// see http://golang.org/src/pkg/sync/atomic/doc.go#L41
+	sess int64
+	// KeepDestinationHeaders indicates the proxy should retain any headers present in the http.Response before proxying
+	KeepDestinationHeaders bool
+	// setting Verbose to true will log information on each request sent to the proxy
+	Verbose         bool
+	Logger          Logger
+	NonproxyHandler http.Handler
+	RoundTripper    *RpRoundTripper
+
+	// ConnectDial will be used to create TCP connections for CONNECT requests
+	ConnectDial func(network string, addr string) (net.Conn, error)
+}
+
+func NewRecorderProxy(port int, conn Connections, connectionTimeout time.Duration) *RecorderProxy {
+	r := &RecorderProxy{
+		id:                proxyCount,
+		conn:              conn,
+		addr:              ":" + strconv.Itoa(port),
+		ConnectionTimeout: connectionTimeout,
+
+		NonproxyHandler: http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			http.Error(w, "This is a proxy server. Does not respond to non-proxy requests.", 500)
+		}),
+		RoundTripper: NewRpRoundTripper(),
+	}
+
+	cache := viper.GetString("cache")
+	if cache != "" {
+		cu, _ := url.Parse("http://" + cache)
+		r.RoundTripper.Proxy = http.ProxyURL(cu)
+		r.ConnectDial = r.NewConnectDialToProxy(cache)
+	}
+
+	r.Logger = log.New(os.Stderr, "", log.LstdFlags)
```

Comment:
Should every recorderProxy create a new logger? Maybe a shared logger should be created in `main` function and be an argument to `NewRecorderProxy`?

Reply:
Won't fix for now